tesseract_time_parameterization/src/time_optimal_trajectory_generation.cpp

```diff
@@ -1093,137 +1090,105 @@ bool Trajectory::isValid() const { return valid_; }
+  double length = 0;
+  for (Eigen::Index i = 1; i < trajectory.size(); ++i)
+  {
+    length += (trajectory.getPosition(i) - trajectory.getPosition(i - 1)).norm();
+    path_lengths[i] = length;
+  }
+
+  if (tesseract_common::almostEqualRelativeAndAbs(length, 0))
+    throw std::runtime_error("Tried to assign trajectory with length zero.");
+
+  double scale = path_.getLength() / length;
+  path_lengths = scale * path_lengths;
+
   long cnt = 0;
-  double error = 0;
-  double prev_error = 1;
-  bool hit_tolerance = false;
-  bool error_increasing = false;
+  Eigen::Index sample_size = std::max(static_cast<Eigen::Index>(getDuration() / 0.01), trajectory.size());
+  Eigen::VectorXd times = Eigen::VectorXd::LinSpaced(sample_size, 0, getDuration());
   for (long i = 0; i < trajectory.size(); ++i)
   {
-    const Eigen::VectorXd& p = trajectory.getPosition(i);
+    Eigen::VectorXd uv;
+    Eigen::VectorXd ua;
+    double time{ 0 };
 
     if (i == 0)
     {
-      // Set first point
-      Eigen::VectorXd uv = getVelocity(trajectory_.front().time_).head(trajectory.dof());
-      Eigen::VectorXd ua = getAcceleration(trajectory_.front().time_).head(trajectory.dof());
-      trajectory.setData(0, uv, ua, it->time_);
-      ++cnt;
+      uv = getVelocity(0).head(trajectory.dof());
+      ua = getAcceleration(0).head(trajectory.dof());
+      time = 0;
+      cnt++;
     }
-    else if (i == (trajectory.size() - 1))
+    else if (i == trajectory.size() - 1)
     {
-      // Set last point
-      Eigen::VectorXd uv = getVelocity(trajectory_.back().time_).head(trajectory.dof());
-      Eigen::VectorXd ua = getAcceleration(trajectory_.back().time_).head(trajectory.dof());
-      trajectory.setData(trajectory.size() - 1, uv, ua, it->time_);
-      ++cnt;
+      time = getDuration();
+      uv = getVelocity(time).head(trajectory.dof());
+      ua = getAcceleration(time).head(trajectory.dof());
+      cnt++;
     }
     else
     {
-      for (; it != trajectory_.end(); ++it)
+      const Eigen::VectorXd& p = trajectory.getPosition(i);
+      double path_length = path_lengths[i];
+      // calculate distance from this point to all other points in the trajectory
+
+      Eigen::VectorXd x(sample_size);
+      Eigen::VectorXd dists(sample_size);
+      for (Eigen::Index j = 0; j < sample_size; ++j)
       {
-        Eigen::VectorXd compare_p = getPosition(it->time_).head(trajectory.dof());
-        error = (p - compare_p).norm();
+        PathData data = getPathData(times[j]);
+        x[j] = data.path_pos;
+        dists[j] = (getPosition(times[j]).head(trajectory.dof()) - p).norm();
+      }
 
-        // Check if we've hit the tolerance and if the error is still decreasing
-        if (error < path_tolerance)
-        {
-          hit_tolerance = true;
-        }
-        if (prev_error < error)
+      // calculate the numerical diff
+      Eigen::VectorXd ndiff(sample_size - 1);
+      for (Eigen::Index i = 0; i < sample_size - 1; ++i)
```

Comment:
I think this either needs to be `sample_size - 2` or else `ndiff.size() - 1` doesn't it? Otherwise `minimums.push_back(i + 1);` could be outside the range of x

Reply:
The index is related to the original data `dist` so it should not be outside the range.